bin/common.ml

```diff
@@ -838,7 +838,9 @@ module Builder = struct
         { x
         ; profile
         ; instrument_with
-        ; workspace_file = Option.map workspace_file ~f:Arg.Path.path
+        ; workspace_file =
+            Option.map workspace_file ~f:(fun p ->
+                Path.Outside_build_dir.External (Arg.Path.External.path p))
```

Comment:
At first glance, it seems that the unsafety just moved here. Can't we construct a non-sensical path here if we start with garbage?

Reply:
The returned path is always "external". It is interpreted as:

* An absolute path if it starts with a `/`
* A relative path to the initial cwd if it doesn't start with a `/`.

So there's no need for validation.

I suppose this could mean that the original `Arg.Path.t` could always return `Path.External.t`. But this is the one path where the distinction actually causes theses _exn conversions.

Reply:
> * A relative path to the initial cwd if it doesn't start with a `/`.

And what if that relative path is `_build/default/some/path`? Then it won't be outside the build dir.

Reply:
Yes, but it will still work the same way as before the patch. It would be treated as an external path regardless.

Reply:
Ah, so you're saying that `Path.as_outside_build_dir_exn` isn't doing any validation either? That seems sad.

Maybe we should fix this problem while we're cleaning this part of the world. I mean, we can at least validate things here, where broken paths can sneak into Dune from the CLI.

Reply:
(Otherwise, I don't see the point of the `Path.Outside_build_dir` abstraction if it doesn't actually provide any guarantees.)

Reply:
The particular issue here is that it's actually quite hard to tell if a path is inside the build directory in the presence of symlinks. For example, one can run things with `--build-dir /foo/bar/build` and then have a symlink to `/foo`. Therefore we just give up and don't try very hard. I don't think it's so bad. At the end of the day, the important thing here is that we aren't doing any operations on this path that are only allowed on build targets.

Reply:
So, to sum up, this PR fixes a superficial problem but leaves a deep problem unsolved :) So be it. 

But maybe worth writing this up in a comment/issue, since this an obvious problem with the current state of things.

Reply:
Yes indeed. The way i see it, we should at least not lie to the user and pretend that we can return a build path.